src/common/utils/query/visualization-query.ts

```diff
@@ -111,12 +112,31 @@ export default function makeQuery(essence: Essence, timekeeper: Timekeeper): Exp
 
   const timeShiftEnv: TimeShiftEnv = essence.getTimeShiftEnv(timekeeper);
 
-  const mainExp: Expression = ply().apply("main", $main.filter(mainFilter.toExpression(dataCube)));
+  let millisecondsInInterval = 1;
+
+  const currentTimeFilter = essence.currentTimeFilter(timekeeper);
+  if (!currentTimeFilter.values.isEmpty()) {
+    const { start, end } = currentTimeFilter.values.get(0);
+    const currentTimeRange = new Duration(start, end, essence.timezone);
+    millisecondsInInterval = currentTimeRange.getCanonicalLength();
+  }
+
+  const mainExp: Expression = ply()
+    .apply("main", $main.filter(mainFilter.toExpression(dataCube)))
+    .apply("MillisecondsInInterval", millisecondsInInterval);
 
   const queryWithMeasures = applySeries(essence.getConcreteSeries(), timeShiftEnv)(mainExp);
 
   if (splits.length() > 0) {
-    return queryWithMeasures.apply(SPLIT, applySplit(0, essence, timeShiftEnv));
+    splits.splits.forEach(split => {
```

Comment:
> I'm not sure if I follow.
> 
> Why do we need to search all splits?

Order of splits is not fixed in table view, but we may be able to stop searching when one is found, see below.

> And is it possible that we will get multiple good values here?

Probably not.

> Why last one is privileged?

There should only be one effective time split. Even if it's split twice, the last one would become the granularity of the final result (e.g. split by 1 hour and then by 5 minutes).

> Maybe it should be `find` instead of `forEach`?

If there's only one effective time split, yes.

> And do we wan't to check if bucket is Duration or if it's really a split on time (the "time", defined in dataCube) dimension?

Good question, does Druid allow two time dimensions in one datasource?

Reply:
@mkuthan and @piter75 correct me if I'm wrong but I think Druid allows only one "__time" filter which describe time of event, but allows additional dimensions which hold datetime values. I think we should search for this "__time" and only then change this milliseconds value.